enhancements/cloud-integration/openstack/cloud-provider-config-upgrade.md

```diff
@@ -0,0 +1,327 @@
+providers are similar, but not identical.
+
+This enhancement will describe how we can ensure correct configuration is passed
+to the cloud provider in use during and after an upgrade, and how we can prevent
+an upgrade to 4.11 if the configuration cannot be upgraded.
+
+## Motivation
+
+The legacy cloud provider is dead. Long live the external cloud provider.
+
+### Goals
+
+- No user action is required in the common case
+- Upgrade legacy cloud provider to external cloud controller manager where possible
+- Mark cluster as non-upgradable where upgrade to external CCM is not possible
+
+### Non-Goals
+
+- Modify user-managed configuration
+- Manage cloud providers for other, non-OpenStack platforms
+
+## Proposal
+
+- The change will be implemented in CCO
+- CCO will create a separate ConfigMap for the external cloud provider
+- We will prevent upgrade if it is not possible to create the external cloud provider ConfigMap
+
+### Implemented in cluster config operator (CCO)
+
+The new functionality will be implemented as a change to CCO. CCO already
+applies a transformation function to the user-managed config and writes the
+result to `openshift-config-managed/kube-cloud-config`. The transformation
+function is selected according to PlatformType in the cluster Infrastructure
+object. Components requiring legacy cloud provider configuration, including
+kubelet and KCM, read the transformed ConfigMap.
+
+### Separate ConfigMaps for legacy and external cloud providers
+
+The cloud configuration is read by multiple components which will not all be
+upgraded simultaneously. If we simply updated the `kube-cloud-config` to be
+compatible with external CCM we would risk breaking KCM and kubelets which
+have not yet been upgraded. Therefore we are constrained to write two versions
+of the managed config map.
+
+CCO will be modified to create both `kube-cloud-config` and
+`kube-cloud-config-external`. `kube-cloud-config` will remain compatible with
+the legacy cloud provider. `kube-cloud-config-external` will be compatible with
+the external cloud provider.
+
+The external CCM will read its configuration from `kube-cloud-config-external`.
+
+### Preventing upgrade when config cannot be upgraded
+
+If CCO is not able to write `kube-cloud-config-external` it will mark itself as
+non-upgradable with a user-actionable error message explaining how
+`openshift-config/cloud-provider-config` is incompatible.
+
+As we intend to switch to external CCM in 4.11, this change will be implemented
+in OCP 4.10.
+
+### Compatibility constraints on user-managed config
+
+This has implications for the user-managed configuration:
+
+In 4.10, the user-managed configuration **must** be compatible with the legacy
+cloud provider. The user **must not** specify config options which cannot be
+transformed to config for the external cloud provider. If either of these are
+not satisfied we will mark CCO as not upgradeable.
```

Comment:
Is there feature parity between the CCMs? Are we going to block some user upgrading because they have some unsupported option? Will they be stuck at 4.10 forever?

It's not obvious to me (yet) how a user fixes issues that arise here